Add: ProductFilters Services Provider

dinhtungdu opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Partially close #43493.

This PR refactors the Blocks\QueryFilters class into several services under the Core namespace. This PR adds the new services and updates new filter blocks to use the new services. Later on, there are several follow up PRs:
- Add cache to boost the performance.
- Deprecate the legacy class.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

This can be tested by developers only.

0. Ensure the experimental blocks are enabled:
    - If you use the WooCommerce Beta Tester plugin, you can enable the experimental blocks in Tools > WCA Test Helper > Features. Toggle the `experimental-blocks` option.
    - If you build this PR locally, build the branch with `pnpm --filter='@woocommerce/plugin-woocommerce' watch:build` command to have experimental blocks available.
1. Add `Product Filters` blocks to the Product Catalog template.
2. See the filter blocks work exactly as they were before, same behavior and product counts.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

dinhtungdu committed:
feat: product filters service provider

dinhtungdu committed:
test: add unit tests

dinhtungdu committed:
chore: changelog

dinhtungdu committed:
chore: lint

dinhtungdu committed:
refactor: rename Counts to FilterData

dinhtungdu committed:
fix: use correct controller class

dinhtungdu committed:
fix: rename pre filter

dinhtungdu committed:
fix: unit test

Aljullu commented on plugins/woocommerce/package.json:
Are the changes in this file related to this PR?

dinhtungdu replied:
No, it's my mistake. Thanks for the catch.

Aljullu commented on plugins/woocommerce/src/Internal/DependencyManagement/ServiceProviders/ProductFiltersServiceProvider.php:
```diff
@@ -0,0 +1,40 @@
+use Automattic\WooCommerce\Internal\ProductFilters\QueryClauses;
+
+/**
+ * LoggingServiceProvider class.
```
This comment needs to be updated, no? :thinking:

dinhtungdu replied:
Yes, my bad. Thanks.

Aljullu commented on plugins/woocommerce/src/Internal/ProductFilters/QueryClauses.php:
```diff
@@ -0,0 +1,401 @@
+	 * @return array
+	 */
+	public function add_query_clauses( $args, $wp_query ) {
+		// Rating filter is handled by tax query.
```
I'm a bit confused by this comment. It says the "Rating filter is handled by tax query", but the logic refers to the stock filter. Do you think the comment could be update/expanded or am I missing something obvious? :sweat_smile:

dinhtungdu replied:
I updated the comment to add more details.

Aljullu replied:
Much clearer, thanks!

Aljullu commented on plugins/woocommerce/src/Internal/ProductFilters/QueryClauses.php:
```diff
@@ -0,0 +1,401 @@
+	 * @return array
+	 */
+	public function add_query_clauses( $args, $wp_query ) {
+		// Rating filter is handled by tax query.
+		if ( $wp_query->get( 'filter_stock_status' ) ) {
+			$stock_statuses = trim( $wp_query->get( 'filter_stock_status' ) );
+			$stock_statuses = explode( ',', $stock_statuses );
+			$stock_statuses = array_filter( $stock_statuses );
+
+			$args = $this->add_stock_clauses( $args, $stock_statuses );
```
Apparently, every time we call `add_stock_statuses` we need to call `trim()`, `explode()` and `array_filter()` before. Wondering if you considered moving those three lines inside the `add_stock_statuses` method.

dinhtungdu replied:
We can move the `array_filter` call only, as the `add_stock_statuses` method accepts an array of statuses.

Aljullu replied:
I guess we can change the method signature given that it's a new class, no? But yeah, not a strong opinion. In fact, I agree that it feels cleaner to pass an array instead of a string that will need to be exploded. So feel free to leave it as it is now. :slightly_smiling_face:

dinhtungdu replied:
>  In fact, I agree that it feels cleaner to pass an array instead of a string that will need to be exploded.

Yeah, that's what I think, too. Passing an array is more explicit and clearer than passing a comma-separated string.  I don't want to lock the implementation for a specific use case. This class is meant to use not only Product Filters but also Product Collection, so it should be generic rather than specific.

Aljullu commented on plugins/woocommerce/src/Internal/ProductFilters/QueryClauses.php:
```diff
@@ -0,0 +1,401 @@
+		);
+
+		$args['join']   = $this->append_product_sorting_table_join( $args['join'] );
+		$args['where'] .= ' AND wc_product_meta_lookup.stock_status IN ("' . implode( '","', $stock_statuses ) . '")';
```
Could we call `esc_sql` here instead of in line 69? It shouldn't make a big difference, but I think the closer to the query we escape the better, no?

dinhtungdu replied:
Can you share an example of how you would call it? It only makes sense to use `esc_sql` for quoted value, otherwise, the query is still vulnerable.

Aljullu replied:
I thought `esc_sql( implode( '","', $stock_statuses ) )` would work? Even though the [docs](https://developer.wordpress.org/reference/functions/esc_sql/) say "It will only escape values to be used in strings in the query", that's exactly what's happening here, no? We only need to escape `$stock_statuses`, which will be between quotes. :thinking:

dinhtungdu replied:
We need to pass multiple values (string) to `IN`, not a comma-separate string, which is only one value.

Aljullu replied:
I'm not sure to understand why `esc_sql( implode( '","', $stock_statuses ) )` wouldn't work. :sweat_smile: But it was a small nit, so I'm happy to leave it as-is. Definitely not a blocker.

Aljullu commented on plugins/woocommerce/src/Internal/ProductFilters/FilterData.php:
```diff
@@ -0,0 +1,349 @@
+				FROM {$wpdb->posts} as posts
+				INNER JOIN {$wpdb->postmeta} as postmeta ON posts.ID = postmeta.post_id
+				AND postmeta.meta_key = '_stock_status'
+				AND postmeta.meta_value = '{$status}'
```
Same as a previous comment, I think it would be clearer to have the `esc_sql` here.

dinhtungdu replied:
Yeah, fixed in [`bd3b790` (#52403)](https://github.com/woocommerce/woocommerce/pull/52403/commits/bd3b7900ea4564f0e9c2c6e0670f3871f2f01086)

Aljullu replied:
Do you know what happened with this? Could we run `esc_sql` directly here?

dinhtungdu replied:
Look like it's lost in a rebase.

dinhtungdu replied:
Addressed in fd4f6a233536.

Aljullu commented on plugins/woocommerce/src/Internal/ProductFilters/FilterData.php:
```diff
@@ -0,0 +1,349 @@
+				AND postmeta.meta_key = '_stock_status'
+				AND postmeta.meta_value = '{$status}'
+				WHERE posts.ID IN ( {$product_query_sql} )
+			";
```
No need to fix it now, but just keeping in mind that if we ever split Product storage out from the posts table, similarly to what was done with orders (HPOS), this kind of code will fail, no?

dinhtungdu replied:
Yes, it definitely will. The development for the custom product table has been pausing for three years, so I guess it won't happening soon. Even if it will, then we can update it later, along with other parts of our codebase that relies on products in post table.

Aljullu commented on plugins/woocommerce/src/Internal/ProductFilters/README.md:
```diff
@@ -0,0 +1,34 @@
+
+Current filter blocks and former filter widgets only care about the product catalog query or the query that returns all products. For example, even when added to the Hoodies category, the filter blocks still show data for all products, including ones belonging to the T-shirts category.
+
+Before Gutenberg, we have product archive contexts including Catalog, Taxonomy, Attribute, and Search. With blocks, we have StoreAPI and Product CollectionBoth. All types of context can be customized by altering the product archive query, passing more parameters, or adding Product Collection backend filters.
```
`CollectionBoth` seems like a typo. :eyes:

Aljullu commented on plugins/woocommerce/src/Internal/ProductFilters/README.md:
:clap:

Aljullu reviewed (commented):
This PR is testing well, and the code looks good to me. I just left some minor comments inline.

Having said that, most of this code and logic was new to me, so I would appreciate it if somebody else could do a second review.

thealexandrelara commented on plugins/woocommerce/src/Internal/ProductFilters/FilterData.php:
```diff
@@ -0,0 +1,349 @@
+		remove_filter( 'posts_clauses', array( $this->query_clauses, 'add_query_clauses' ), 10 );
+		remove_filter( 'posts_pre_query', '__return_empty_array' );
+
+		$price_filter_sql = "
+		SELECT min( min_price ) as min_price, MAX( max_price ) as max_price
+		FROM {$wpdb->wc_product_meta_lookup}
+		WHERE product_id IN ( {$product_query_sql} )
+		";
```
Should it be using `$wpdb->prepare`? 

```suggestion
		$price_filter_sql = $wpdb->prepare(
                "
                    SELECT MIN(min_price) as min_price, MAX(max_price) as max_price
                    FROM {$wpdb->wc_product_meta_lookup}
                    WHERE product_id IN ( %s )
                ",
                $product_query_sql
);
```

dinhtungdu replied:
We can't. `$product_query_sql` is already prepared. Running through another preparation causes the query to return nothing.
<img width="1512" alt="image" src="https://github.com/user-attachments/assets/2ec92e83-2f8d-48aa-a6e0-cf05ce1231c3">

dinhtungdu replied:
@thealexandrelara We can't. `$product_query_sql` is already prepared. Running through another prepare causes the query to return nothing. Same for your other comments, too.

<img width="1512" alt="image" src="https://github.com/user-attachments/assets/2ec92e83-2f8d-48aa-a6e0-cf05ce1231c3">

thealexandrelara commented on plugins/woocommerce/src/Internal/ProductFilters/FilterData.php:
```diff
@@ -0,0 +1,349 @@
+			AND average_rating > 0
+			GROUP BY rounded_average_rating
+			ORDER BY rounded_average_rating DESC
+		";
```
I'm wondering if here it would be the case to also use `$wpdb->prepare`:

```suggestion		
                $rating_count_sql = $wpdb->prepare(
			"
			SELECT COUNT( DISTINCT product_id ) as product_count, ROUND( average_rating, 0 ) as rounded_average_rating
			FROM {$wpdb->wc_product_meta_lookup}
			WHERE product_id IN ( %s )
			AND average_rating > 0
			GROUP BY rounded_average_rating
			ORDER BY rounded_average_rating DESC
			",
			$product_query_sql
		);
```

thealexandrelara commented on plugins/woocommerce/src/Internal/ProductFilters/FilterData.php:
```diff
@@ -0,0 +1,349 @@
+			WHERE posts.ID IN ( {$product_query_sql} )
+			{$attributes_to_count_sql}
+			GROUP BY terms.term_id
+		";
```
Same here:

```suggestion
		$attribute_count_sql = $wpdb->prepare(
			"
			SELECT COUNT( DISTINCT posts.ID ) as term_count, terms.term_id as term_count_id
			FROM {$wpdb->posts} AS posts
			INNER JOIN {$wpdb->term_relationships} AS term_relationships ON posts.ID = term_relationships.object_id
			INNER JOIN {$wpdb->term_taxonomy} AS term_taxonomy USING( term_taxonomy_id )
			INNER JOIN {$wpdb->terms} AS terms USING( term_id )
			WHERE posts.ID IN ( %s )
			%s
			GROUP BY terms.term_id
			",
			$product_query_sql,
			$attributes_to_count_sql
		);
```

thealexandrelara reviewed (dismissed):
I’ve left some comments on the code. I haven’t worked on this part before, so I’ll ask someone else to take another look as well. Overall, I didn’t find any issues during testing, so I’m approving the PR! ✅ Great work!

Here are the testing results:

✅ Scenario 1: The Product Filters (Experimental) blocks works in an empty store (without products, attributes, etc).
- Conditions:
    - A store does not have any products, attributes, and categories registered.
    - A merchant adds the Product Filters (Experimental) block to the Product Catalog template.
    - A shopper visits the Shop page.
- Expected Outcome: No error should appear on the frontend.
- Summary: Scenario passed as expected. No issues were found.

---

✅ Scenario 2: The Product Filters (Experimental) blocks correctly renders when the store has products and attributes.
- Conditions:
    - A store has products, attributes, and categories registered.
    - A merchant adds the Product Filters (Experimental) block to the Product Catalog template.
    - A shopper visits the Shop page.
- Expected Outcome: The filters are rendered as expected.
- Summary: Scenario passed as expected. No issues were found.

---

✅ Scenario 3: Shoppers can modify the price filter values.
- Conditions:
    - A store has products, attributes, and categories registered.
    - A merchant adds the Product Filters (Experimental) block to the Product Catalog template.
    - A shopper visits the Shop page.
    - The shopper modifies the price filter to different values.
- Expected Outcome: 
    - The Active Filters block displays the selected range.
    - The Price Filter updates with the selected values. 
    - Both min and max values are updated when shoppers interact with the slider. 
    - The URL updates with the `max_price` and `min_price` query params.
    - The clear button works as expected and resets the filter state to its initial value.
    - When fully reloading the page with values set to the Price Filter, the block is pre-populated correctly.
- Summary: Scenario passed as expected. No issues were found.

---

✅ Scenario 4: Shoppers can modify the Rating Filter values.
- Conditions:
    - A store has products, attributes, and categories registered.
    - A merchant adds the Product Filters (Experimental) block to the Product Catalog template.
    - A shopper visits the Shop page.
    - The shopper selects some filter options for the Rating Filter.
- Expected Outcome: 
    - The Active Filters block displays the selected filter options.
    - The Rating Filter updates with the selected values. 
    - The URL updates with the `filter_rating` query param.
    - The clear button works as expected and resets the filter state to its initial value.
    - When fully reloading the page with filter options selected for the Rating Filter, the block is pre-populated correctly.
- Summary: Scenario passed as expected. No issues were found.

---

✅ Scenario 5: Shoppers can modify the Attribute Filter values.
- Conditions:
    - A store has products, attributes, and categories registered.
    - A merchant adds the Product Filters (Experimental) block to the Product Catalog template.
    - A shopper visits the Shop page.
    - The shopper selects some filter options for the Attribute Filter.
- Expected Outcome: 
    - The Active Filters block displays the selected filter options.
    - The Attribute Filter updates with the selected values. 
    - The URL updates with the attribute filter query params.
    - The clear button works as expected and resets the filter state to its initial value.
    - The number of products displayed matches the products count.
    - When fully reloading the page with filter options selected for the Attribute Filter, the block is pre-populated correctly.
- Summary: Scenario passed as expected. No issues were found.

---

✅ Scenario 6: Shoppers can modify the Status Filter values.
- Conditions:
    - A store has products, attributes, and categories registered.
    - A merchant adds the Product Filters (Experimental) block to the Product Catalog template.
    - A shopper visits the Shop page.
    - The shopper selects some filter options for the Status Filter.
- Expected Outcome: 
    - The Active Filters block displays the selected filter options.
    - The Status Filter updates with the selected values. 
    - The URL updates with the `filter_stock_status` query param.
    - The clear button works as expected and resets the filter state to its initial value.
    - The number of products displayed matches the products count.
    - When fully reloading the page with filter options selected for the Status Filter, the block is pre-populated correctly.
- Summary: Scenario passed as expected. No issues were found.

---

✅ Scenario 7: Shoppers can combine multiple filters.
- Conditions:
    - A store has products, attributes, and categories registered.
    - A merchant adds the Product Filters (Experimental) block to the Product Catalog template.
    - A shopper visits the Shop page.
    - The shopper selects some filter options for multiple filters.
- Expected Outcome: 
    - The Active Filters block updates with the selected filters options.
    - The URL updates with the selected filters query params.
    - The clear button works as expected and resets the filter state to its initial value.
    - The number of products displayed matches the products count.
    - When fully reloading the page with filter options selected for all filters, they are pre-populated correctly.
- Summary: Scenario passed as expected. No issues were found.

dinhtungdu committed:
chore: reset package.json

dinhtungdu commented on plugins/woocommerce/src/Internal/DependencyManagement/ServiceProviders/ProductFiltersServiceProvider.php:
```diff
@@ -0,0 +1,40 @@
+	public function register() {
+		$this->share( QueryClauses::class );
+		$this->share_with_implements_tags( Controller::class )->addArgument( QueryClauses::class );
+		/**
+		 * We allow changing the clauses generator at run time, so we use `add`
+		 * here to return a new instance with a known default clause generator
+		 * when retrieving the data provider from the container.
+		 */
+		$this->add( FilterData::class )->addArgument( QueryClauses::class );
```
We need to change the design of this class because of the new container added in [#52296](https://github.com/woocommerce/woocommerce/pull/52296) which [doesn't support factory](https://github.com/woocommerce/woocommerce/blob/4214c5c675883382583b29370f7bcc219994ad17/plugins/woocommerce/src/Internal/DependencyManagement/RuntimeContainer.php) at the moment:

dinhtungdu committed:
fix: integrate with new RuntimeContainer

dinhtungdu committed:
chore: fix service name

dinhtungdu committed:
docs: typo

dinhtungdu committed:
refactor: rename setter

dinhtungdu committed:
chore: explain the reason for no rating clause

dinhtungdu committed:
refactor: clean up the stock array before process

dinhtungdu committed:
refactor: escaping as late as possible

dinhtungdu committed:
refactor: use a provider to init, cache, and retrieve FilterData instance with desired query clauses generator

dinhtungdu committed:
refactor: required passing stock statuses as second argument

dinhtungdu commented:
Hi there, in [`18973ac` (#52403)](https://github.com/woocommerce/woocommerce/pull/52403/commits/18973ac9a558dfa95b61241a0686b45364b3fd3d), I updated the design of the new service to remove the need of a factory, so it's compatible with the new (now merged) `RuntimeContainer`, as it's an important change, can you please take another look at this? cc @Aljullu @thealexandrelara

kalessil commented:
@dinhtungdu: quick question on the requested review from Flux - Is there any specific focus for our review?

dinhtungdu committed:
chore: post rebase fixes

dinhtungdu committed:
fix: issue with php 7.4 where dynamic class :class is not available

dinhtungdu committed:
test: update unit test in favor of the new design

dinhtungdu committed:
chore: phpcs fix

roykho reviewed (approved):
Great work! In the notes, you mentioned you will deprecate the legacy class. Where did you do that? Other than that I think this PR is good to go.

✅ Smoke tested the filters
✅ Tested with different counts/order settings

dinhtungdu commented:
@roykho thanks for the reminder, I will do that in a follow up PR as this one is already so big.

dinhtungdu commented:
> quick question on the requested review from Flux - Is there any specific focus for our review?

@kalessil yes, it's code-owner-related and auto assignment. But I'd love to have a review from Flux.

Aljullu commented on plugins/woocommerce/src/Internal/ProductFilters/Controller.php:
```diff
@@ -0,0 +1,85 @@
+<?php
+/**
+ * MainQueryFilters class file.
```
Is this comment correct? I even think we could remove it, no? It doesn't seem to add much information.

dinhtungdu replied:
This is left over from the previous iteration. I will update it with the correct information. Thanks!

dinhtungdu replied:
I removed the file comment and keep only class comment.

Aljullu commented on plugins/woocommerce/src/Internal/ProductFilters/Controller.php:
```diff
@@ -0,0 +1,85 @@
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * MainQueryFilters class.
```
And this?

Aljullu commented on plugins/woocommerce/src/Internal/ProductFilters/Controller.php:
```diff
@@ -0,0 +1,85 @@
+	 * Hook into actions and filters.
+	 */
+	public function register() {
+		self::add_filter( 'posts_clauses', array( $this, 'main_query_filter' ), 10, 2 );
+		self::add_action( 'woocommerce_after_product_object_save', array( $this, 'clear_filter_data_cache' ) );
+		self::add_action( 'woocommerce_delete_product_transients', array( $this, 'clear_filter_data_cache' ) );
```
I wonder if we could divide this into two classes, one that runs in the frontend for the `post_clauses` filter and another one that runs in the admin for the actions. Would that work?

dinhtungdu replied:
Makes sense! Let me think more about this.

dinhtungdu replied:
In ce135d91f41d, I split that class into 2 one with separate responsibilities.

Aljullu commented on plugins/woocommerce/tests/php/src/Internal/ProductFilters/QueryClausesTest.php:
```diff
@@ -0,0 +1,192 @@
+	 *           ["pa_color",["red-slug","not-exist-slug"],"or"]
+	 *           ["pa_color",["red-slug","green-slug"],"or"]
+	 *
+	 * Skipping tests with query_type `and` because there is an issue with
```
Should we create a follow-up issue for this? Or maybe add a `@todo` in the comment?

dinhtungdu replied:
Let's add a todo.

Aljullu commented on plugins/woocommerce/tests/php/src/Internal/ProductFilters/QueryClausesTest.php:
```diff
@@ -0,0 +1,192 @@
+	 * for `and`. We can fix the same issue in
+	 * FilterClausesGenerator::add_attribute_clauses but doing so will make the
+	 * attribute counts data doesnt match with current query. A fix for both
+	 * methods is pending.
```
Small suggestion:

```suggestion
	 * Filterer::filter_by_attribute_post_clauses that generates wrong clauses
	 * for `and`. We can fix the same issue in
	 * FilterClausesGenerator::add_attribute_clauses but doing so will make the
	 * attribute counts data not match the current query. A fix for both
	 * methods is pending.
```

Aljullu commented on plugins/woocommerce/tests/php/src/Internal/ProductFilters/QueryClausesTest.php:
```diff
@@ -0,0 +1,192 @@
+	/**
+	 * The system under test.
+	 *
+	 * @var DataRegenerator
```
Shouldn't this type be `QueryClauses`? :thinking:

dinhtungdu replied:
Yes, look like a copy pasta left over. Thanks!

Aljullu commented on plugins/woocommerce/tests/php/src/Internal/ProductFilters/AbstractProductFiltersTest.php:
```diff
@@ -0,0 +1,329 @@
+			"
+		);
+
+		update_option( 'woocommerce_attribute_lookup_enabled', 'yes' );
+		update_option( 'woocommerce_calc_taxes', 'no' );
+		update_option( 'woocommerce_tax_display_shop', 'excl' );
```
I'm not sure how PHP Unit Tests work, but shouldn't we reset these options after the tests have run?

dinhtungdu replied:
Good point. Let's backup those options.

Aljullu commented on plugins/woocommerce/tests/php/src/Internal/ProductFilters/AbstractProductFiltersTest.php:
```diff
@@ -0,0 +1,329 @@
+		$this->fixture_data = new FixtureData();
+
+		global $wpdb;
+		$wpdb->query(
+			"
+			  CREATE TABLE IF NOT EXISTS {$wpdb->prefix}wc_product_attributes_lookup (
+			  product_id bigint(20) NOT NULL,
+			  product_or_parent_id bigint(20) NOT NULL,
+			  taxonomy varchar(32) NOT NULL,
+			  term_id bigint(20) NOT NULL,
+			  is_variation_attribute tinyint(1) NOT NULL,
+			  in_stock tinyint(1) NOT NULL
+			  );
+			"
+		);
```
Why do we need to create this table manually in the test? Could we add a comment explaining that?

dinhtungdu replied:
Test can be run as a whole or single test suite. If we filter to run only this test, the table may not exist. I will add a comment.

Aljullu replied:
Ok, it makes sense, thanks for adding the comment!

Aljullu commented on plugins/woocommerce/tests/php/src/Internal/ProductFilters/AbstractProductFiltersTest.php:
```diff
@@ -0,0 +1,329 @@
+	}
+
+	/**
+	 * Build and create attribute from variations data.
```
I guess this should be a plural?

```suggestion
	 * Build and create attributes from variations data.
```

Aljullu commented on plugins/woocommerce/tests/php/src/Internal/ProductFilters/AbstractProductFiltersTest.php:
```diff
@@ -0,0 +1,329 @@
+			array(
+				'name'          => 'Product 1',
+				'regular_price' => 10,
+				'stock_status'  => 'onbackorder',
```
Not a big deal, but maybe it would be a good idea to use the Product Status enums from [here](https://github.com/woocommerce/woocommerce/blob/f6a2d0a2814e53dea2115d01613fe9a769a9d401/plugins/woocommerce/src/Enums/ProductStockStatus.php#L10)?

dinhtungdu replied:
Good call! Updated.

Aljullu commented on plugins/woocommerce/tests/php/src/Internal/ProductFilters/FilterDataTest.php:
```diff
@@ -0,0 +1,384 @@
+	 * @param callable $filter_callback Callback passed to filter test products.
+	 */
+	private function get_expected_attribute_counts( $attribute_name, $filter_callback = null ) {
+		$attribute_counts  = array();
+		$_attribute_counts = array();
```
Could we find a better name for these variables?  Something like `$filtered_attribute_counts` or something like that?

dinhtungdu replied:
Updated to more explicit names.

Aljullu commented on plugins/woocommerce/tests/php/src/Internal/ProductFilters/FilterDataTest.php:
```diff
@@ -0,0 +1,384 @@
+		$wp_query   = new \WP_Query( array( 'post_type' => 'product' ) );
+		$query_vars = array_filter( $wp_query->query_vars );
+
+		$actual_rating_counts   = (array) $this->sut->get_rating_counts( $query_vars );
```
Do we need to cast this into an `array`? My understanding is that `get_rating_counts()` already returns an array? This pattern repeats in several other places in this file.

dinhtungdu replied:
I don't know why too. I agree that casting looks unneccessary. Let's remove them.

Aljullu commented on plugins/woocommerce/src/Internal/ProductFilters/FilterData.php:
```diff
@@ -0,0 +1,334 @@
+<?php
+/**
+ * Class for filter counts.
+ */
+
```
Are these comments on top of the file required? If not, I would potentially remove them. It feels like unnecessary duplication with the comment before the class.

dinhtungdu replied:
It's mostly to make PHPCS happy. But looks like it's not necessary anymore, let me remove it.

Aljullu commented on plugins/woocommerce/src/Internal/ProductFilters/FilterData.php:
```diff
@@ -0,0 +1,334 @@
+			return $cached_data;
+		}
+
+		add_filter( 'posts_clauses', array( $this->query_clauses, 'add_query_clauses' ), 10, 2 );
+		add_filter( 'posts_pre_query', '__return_empty_array' );
+
+		$query_vars['no_found_rows']  = true;
+		$query_vars['posts_per_page'] = -1;
+		$query_vars['fields']         = 'ids';
+		$query                        = new \WP_Query();
+
+		$query->query( $query_vars );
+		$product_query_sql = $query->request;
+
+		remove_filter( 'posts_clauses', array( $this->query_clauses, 'add_query_clauses' ), 10 );
+		remove_filter( 'posts_pre_query', '__return_empty_array' );
```
Have you considered extracting this into a private method? It seems to be repeated several times.

dinhtungdu replied:
Yes, I add serveral improvements in https://github.com/woocommerce/woocommerce/pull/52449/

Aljullu commented on plugins/woocommerce/src/Internal/ProductFilters/FilterData.php:
```diff
@@ -0,0 +1,334 @@
+		 * @param array $query_vars   The query arguments to calculate the filter data.
+		 * @param array $extra        Some filter types require extra arguments for calculation, like attribute.
+		 */
+		return apply_filters( 'woocommerce_pre_product_filter_data', null, $filter_type, $query_vars, $extra );
```
This filter is new from this PR? Do you have some more context on why it's needed?

dinhtungdu replied:
I updated the comment to make it clearer.

Aljullu replied:
Sorry, it's still not clear to me. :sweat_smile: I can't find this filter in `trunk` so I guess it's a new one, but the `@since` tag mentions 8.7.0. I'm also a bit wary of introducing new hooks if we hadn't received a request for it or we need it for a specific extension. :thinking:

dinhtungdu replied:
My bad, I didn't update the version here.

> I'm also a bit wary of introducing new hooks if we hadn't received a request for it or we need it for a specific extension. 🤔

We actually got some request to filter the filter result like https://github.com/woocommerce/woocommerce/issues/42631. The main reason for this filter is that if merchant want to use external services for processing filter data like Elasticsearch, which is quite popular for large stores, we will need this filter for those service to hook in early and skip unnecessary calculation.

Aljullu replied:
Thanks for the extra context, @dinhtungdu!

Aljullu commented on plugins/woocommerce/src/Internal/ProductFilters/QueryClauses.php:
```diff
@@ -0,0 +1,404 @@
+	 *
+	 * @return boolean
+	 */
+	private function adjust_price_filters_for_displayed_taxes() {
```
Given that this function doesn't do anything itself and instead returns a boolean, I think it would be better named `should_adjust_price_filters_for_displayed_taxes()` or something along these lines. What do you think?

Aljullu reviewed (commented):
I like the approach taken in this PR and overall it's working well. I noticed a couple of issues, but I don't know whether they are related to this PR or they should be fixed separately:

1. When a filter doesn't have any available option, we still display the heading, is that expected?

[Enregistrament de pantalla del 2025-03-18 11-45-25.webm](https://github.com/user-attachments/assets/df10f99e-0564-433f-b7af-b731cff81b14)

2. Another question: is it expected that the _Filter by attribute_ doesn't have an option to display the counts? Maybe I'm not finding it. :sweat_smile: 

![imatge](https://github.com/user-attachments/assets/2b3fb392-72ac-452a-8d01-3182cb03a041)

Besides that, I left several comments and questions inline in the code. None of them are strong opinions and in some cases it's just questions, so no need to change anything in some of them.

dinhtungdu committed:
docs: update Controller class comment

dinhtungdu committed:
test: backup and reset filter options after each test

dinhtungdu committed:
docs: explain the reason for creating lookup table

dinhtungdu committed:
test: use better variable names

dinhtungdu committed:
docs: apply Albert review suggestion

dinhtungdu committed:
refactor: add should_ prefix for function returning boolean

dinhtungdu committed:
refactor: use stock status enum

dinhtungdu committed:
test: remove unneccessary casting

dinhtungdu committed:
refactor: split Controller class, remove private action/filter

dinhtungdu commented:
@Aljullu I addressed all of your reviews. Can you take another look at this?

> When a filter doesn't have any available option, we still display the heading, is that expected?

It's not. But https://github.com/woocommerce/woocommerce/pull/56256 should fix it.

> Another question: is it expected that the Filter by attribute doesn't have an option to display the counts? Maybe I'm not finding it. 😅

The count option is in the Styles tab, but https://github.com/woocommerce/woocommerce/pull/56415 moved it to the Settings tab.

dinhtungdu committed:
chore: phpcs

Aljullu commented on plugins/woocommerce/src/Internal/ProductFilters/CacheController.php:
```diff
@@ -0,0 +1,33 @@
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Hooks into WordPress and WooCommerce actions and filters to register main query filter and cache invalidation.
```
Nit: in fact, the hooks in this class only target WooCommerce actions, no? And this class now only handles cache invalidation.

dinhtungdu replied:
Oh, I forgot to update these comments, addressed in 224d569. Thank you!

Aljullu commented on plugins/woocommerce/src/Internal/ProductFilters/MainQueryController.php:
```diff
@@ -0,0 +1,69 @@
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Hooks into WordPress and WooCommerce actions and filters to register main query filter and cache invalidation.
```
Similar to my previous comment, this comment can be probably updated to be more specific. Something like:

```suggestion
 * Hooks into a WordPress filter to register main query filter.
```

dinhtungdu replied:
Addressed in 224d569.

Aljullu reviewed (approved):
Thanks for the updates, @dinhtungdu, they look good to me! :raised_hands: 

I left a couple more comments, one about the filter and another one about [adding a `esc_sql()` here](https://github.com/woocommerce/woocommerce/pull/52403#discussion_r1839968122). Besides that, everything looks good, so pre-approving.

nerrad commented on plugins/woocommerce/src/Internal/ProductFilters/FilterData.php:
```diff
@@ -0,0 +1,331 @@
+		WHERE product_id IN ( {$product_query_sql} )
+		";
+
+		$results = $wpdb->get_row( $price_filter_sql ); // phpcs:ignore
```
It'd be good to include a comment why this is ignored, also, does it have to be a global ignore for the line? Are you sure all the query clauses injected in via `posts_clauses` are prepared?

dinhtungdu replied:
I added comments explaining why we don’t use `$wpdb->prepare` for queries in this class in 701d9a706860. You’re right that it doesn’t have to be global; I changed it to `WordPress.DB.PreparedSQL.NotPrepared`.

> Are you sure all the query clauses injected in via posts_clauses are prepared?

No, it’s the plugin’s responsibility to prepare them, so we can’t say that all query clauses are prepared, so the query built by `WP_Query` in this case is not fully prepared (I was wrong about this). We can only ensure query clauses that come from WooCommerce are prepared.

nerrad commented on plugins/woocommerce/src/Internal/ProductFilters/FilterData.php:
```diff
@@ -0,0 +1,331 @@
+		$query                        = new \WP_Query();
+
+		$query->query( $query_vars );
+		$product_query_sql = $query->request;
```
Seems unnecessary to assign this to a variable? You should be able to use `$query->request` directly later?

dinhtungdu replied:
Yes. This is copied over from Store API ProductQueryFilters. I agree that it's unnecessary. We will eventually [replace](https://github.com/woocommerce/woocommerce/blob/c977084fc5395406fc1eb734266bf7b85baacc97/plugins/woocommerce/src/Internal/ProductFilters/FilterData.php#L89) subquery with cached product IDs. I will remove this unnecessary variable assignment and refactor it later.

dinhtungdu replied:
Addressed in 2423957094f3.

nerrad reviewed (commented):
I had a couple of minutes to take a quick look but haven't had a chance to review the whole thing. Decided to publish what I commented on anyway.

dinhtungdu committed:
docs: update controller class comments

dinhtungdu committed:
chore: update filter since version

dinhtungdu committed:
fix: add esc_sql back

dinhtungdu committed:
feat: skip cache if WP_DEBUG

dinhtungdu committed:
refactor: remove unnecessary variable assignment

dinhtungdu committed:
docs: explain why we don't use wpdb->prepare for filter data query

dinhtungdu committed:
chore: phpcs

dinhtungdu committed:
refactor: pre filter data method name

dinhtungdu committed:
Merge branch 'trunk' into add/product-filters-service-provider
